src/meta/src/manager/scale.rs

```diff
@@ -0,0 +1,245 @@
+                    error!("Failed to solve scale task: {}", err);
+                }
+                permit.forget();
+            }
+        });
+
+        scale_manager
+    }
+
+    async fn solve_task(
+        meta_store: Arc<S>,
+        task_queue: Arc<Mutex<VecDeque<ScaleTask>>>,
+    ) -> Result<()> {
+        let mut task_queue_guard = task_queue.lock().await;
+        // Be called after `task_semaphore.acquire()`, so we can use `unwrap()`.
+        let mut task = task_queue_guard.front_mut().unwrap();
+        task.task_status = TaskStatus::Building as i32;
+        task.insert(&*meta_store).await?;
+
+        match task.get_task_type()? {
+            TaskType::Invalid => unreachable!(),
+            TaskType::ScaleIn => {
+                let _hosts = task.get_hosts();
+                // TODO: call some method to scale in
+
+                task.task_status = TaskStatus::Finished as i32;
+                task.insert(&*meta_store).await?;
+                task_queue_guard.pop_front();
+            }
+            TaskType::ScaleOut => {
+                let _hosts = task.get_hosts();
+                let _fragment_parallelism = task.get_fragment_parallelism();
+                // TODO: call some method to scale out
+
+                let task_id = task.task_id;
+                drop(task_queue_guard);
+
+                // TODO: wait before CN starts.
+                // Temporarily replace them with `sleep` in unit tests.
+                #[cfg(test)]
+                tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+
+                task_queue_guard = task_queue.lock().await;
+                task = task_queue_guard.front_mut().unwrap();
+                if task_id == task.task_id {
+                    task.task_status = TaskStatus::Finished as i32;
+                    task.insert(&*meta_store).await?;
+                    task_queue_guard.pop_front();
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn add_scale_task(&self, mut task: ScaleTask) -> Result<ScaleTaskId> {
+        let task_id = self.next_task_id.fetch_add(1, Ordering::Relaxed);
+        task.task_id = task_id;
+        task.task_status = TaskStatus::Pending as i32;
+        task.insert(&*self.meta_store).await?;
+        self.task_queue.lock().await.push_back(task);
+        self.task_semaphore.add_permits(1);
+        Ok(task_id)
+    }
+
+    pub async fn get_task_status(&self, task_id: ScaleTaskId) -> Result<TaskStatus> {
+        ScaleTask::select(&*self.meta_store, &task_id)
+            .await?
+            .map_or(Ok(TaskStatus::NotFound), |task| Ok(task.get_task_status()?))
+    }
+
+    pub async fn abort_task(&self, task_id: ScaleTaskId) -> Result<()> {
+        let mut task_queue_guard = self.task_queue.lock().await;
+        let index = task_queue_guard
+            .iter()
+            .position(|task| task.task_id == task_id);
```

Comment:
Add task status check, task in `Building` status is not abortable.

Reply:
Here I change the meaning of `Building`. We can make sure the task which are beening done can't be aborted because of we have the `MutexGuard` of task_queue.